### ISSUE TITLE
Update Markdown links when attachment is renamed

### ISSUE THREAD

**Issue by tarheels100** [notable/notable#834]
### Feature description

Automatically update link paths in Markdown files when an attachment is renamed.

### Feature motivation

Renaming an attachment file in the OS is automatically caught by Notable, causing the name of the attachment in the note's attachment list to update automatically. However, the markdown links to this attachment in the note do not update, which to me is the more important part.  

The most critical use case is for images that are pasted from the clipboard directly into Notable.  They are assigned a generic time-stamp filename which the user will probably want to modify to something meaningful.  Having to update the filename in the OS and then again in Notable makes this otherwise great feature about as slow as saving the image manually and then linking to it in Notable.

### Bonus
Allow users to supply a custom filename (with the current time-stamp name offered as a default) when an image is pasted.

**Comment by fabiospampinato** [comment]
> Having to update the filename in the OS and then again in Notable makes this otherwise great feature about as slow as saving the image manually and then linking to it in Notable.

Good point 🤔 

I haven't implemented this yet because basically once it's implemented if a file gets renamed you have to check ALL notes if they are linking to that file, unless we store additional metadata for this, perhaps we should just automatically add a file as an attachment to the note if that note is linking to it 🤔

**Comment by fabiospampinato** [comment]
I don't think this is quite the right way to approach the problem, on the contrary the link one writes in the first place should be pretty much "fixed", in order to do that we need to assign a unique ID to each attachment.

**Comment by tarheels100** [comment]
> I don't think this is quite the right way to approach the problem, on the contrary the link one writes in the first place should be pretty much "fixed", in order to do that we need to assign a unique ID to each attachment.

Maybe, but my preference would be to have the option for meaningful filenames/links for attachments.  Imagine you were trying to manage (insert/replace/rearrange) a bunch of attachments for a note and all their names were random UUID numbers.  Overall, I think there are two questions about how this should work:

1. Should the user have the option of manually modifying/editing the automatically generated filename when an image is pasted into a note?  
    - For me, the answer is strongly yes.  Of course, I could just bypass the "paste image" feature and do everything manually through the OS, but direct pasting is convenient.

2. Should Notable update attachment filenames when they are renamed through the OS?  
    - For me, this is a yes.  The thing that I've never understood about how it's currently implemented is why the attachment list automatically updates in response to a rename, but the filename in the note does not.  What is the use case for this?  It's great that Notable keeps track of the filename, but why go through the trouble if the link in the note breaks?  Please don't take this as confrontational, I just don't really get the reasoning.  However, I may very well be in the minority in this view!

**Comment by fabiospampinato** [comment]
> Imagine you were trying to manage (insert/replace/rearrange) a bunch of attachments for a note and all their names were random UUID numbers. 

That wouldn't be nice, but maybe one can have the best of both worlds. I haven't decided on what the best solution here would be, and so far I don't see any solution that I particularly like honestly, but one could append UUIDs to attachments so that instead of having `Foo.png` it would be named `Foo - 00112233-4455-6677-8899-aabbccddeeff.png` instead.

It wouldn't be a great solution IMHO, but I think we need to be able to assign a unique number to each attachment in order to reliably link to it and detect its changes.

We could add a custom EXIF field for images, but not all file formats have support for custom metadata to be stored about them.

We could instead store their hashes, but if for example an image attachment gets minified or something then we would lose track of it.

We could keep track of hashes _and_ filenames, so that if the hash changes we could update it as long as the filenames stays the same for long enough for the app to detect the change, but things would get tricky, what if the user changes filenames and hashes of 2+ files while the app is closed?

If you have any suggestions I'm all ears.

Also while a link to a UUID might look like this: `@attachment/00112233-4455-6677-8899-aabbccddeeff` the editor may show a tooltip with more clear information about it, or it may even show those links as if they were pointing to `@attachment/Foo.png`, for readability. 

> Should the user have the option of manually modifying/editing the automatically generated filename when an image is pasted into a note?

Yes, but maybe with the caveat that if we will append UUIDs to filenames those shouldn't be touched.

> The thing that I've never understood about how it's currently implemented is why the attachment list automatically updates in response to a rename, but the filename in the note does not.

It's just the default of how things work: we are keeping track of the "attachments" metadata field, so we can update it easily when we detect a change, but in order to update links to attachments inside notes reliably we would have to parse all notes' content, find links pointing to the attachment that changed, and update that, it's a lot more complicate and resource intensive.

Maybe a step in the right direction here would be to automatically keep track of links to attachments as well, perhaps effectively storing a reference in the `attachments` metadata field to each attachment linked automatically. This has been discussed here I believe: #318.

**Comment by bluebirch** [comment]
In the meantime, this simple shellscript can rename a single attachment all all notes linking to it:

```sh
#!/bin/sh

NOTABLE_DIR=$HOME/.notable

if [ -z "$1" -o -z "$2" ] ; then
    echo "Usage: $0 <attachment file name> <new file name>"
fi

if [ -f "$NOTABLE_DIR/attachments/$1" ] ; then
    if [ -f "$NOTABLE_DIR/attachments/$2" ] ; then
        echo "Attachment $2 already exists."
    else
        grep -lZ "$1" $NOTABLE_DIR/notes/*.md | xargs -0 sed -i -e "s/^\\(attachments:.*\\)$1/\\1$2/; s/\(@attachment\/\)$1/\\1$2/"
        mv "$NOTABLE_DIR/attachments/$1" "$NOTABLE_DIR/attachments/$2"
    fi
else
    echo "Attachment $1 does not exist."
fi
```
